test/optimizer/test_memo.py

```diff
@@ -0,0 +1,43 @@
+import unittest
+
+from mock import MagicMock
+
+from src.optimizer.memo import Memo
+from src.optimizer.group import INVALID_GROUP_ID
+
+
+class MemoTest(unittest.TestCase):
+    def test_memo_add_with_no_forcing_id(self):
+        group_expr = MagicMock()
+        group_expr.group_id = INVALID_GROUP_ID
+        memo = Memo()
+        self.assertEqual(memo.add_group_expr(group_expr), group_expr)
+
+    def test_memo_add_with_forcing_id(self):
```

Comment:
@xzdandy @gaurav274 
What is the expected behavior for memo if the program creates group expression with forced id when the memo is empty?
Currently, the program cannot handle this testcase due to index out of bound error.

Reply:
good point, add an assert in b7df722

Reply:
I see. So program should not force id if it is empty? 

I will skip some of my test cases then.

Reply:
Yes. The only time we want to force a id for a group expression is when rewrite in optimizer happens. Say after rewrite, operator **x** become operator **y**. Then the group expression containing operator **y** has to have the same id as the group expression containing operator **x** because operator  **x** can be a child of some other operators and we need to maintain the tree structure after the rewrite.